astropy/utils/decorators.py

```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
                 all_args[arg_type] = None
 
     return all_args
+
+
+def format_doc(docstring, *args, **kwargs):
+    """
+    Replaces the docstring of the decorated object and then formats it.
+
+    This decorator keeps the objects signature and only alters the saved
+    ``__doc__`` property (so it cannot be used as class decorator in python2).
+    The formatting works like :meth:`str.format` and if the decorated object
+    already has a docstring this docstring can be included in the new
+    documentation if you use the ``{__doc__}`` placeholder.
+    It's primary use is if multiple functions have the same or only
+    a slightly different *long* docstring and you want to DRY
+    (https://de.wikipedia.org/wiki/Don%E2%80%99t_repeat_yourself).
+
+    Parameters
+    ----------
+    docstring : str or object
+        The docstring that will replace the docstring of the decorated
+        object. If it is an object like a function or class it will
+        take the docstring of this object. If it is a string it will use the
+        string itself. One special case is if the string is ``'__doc__'`` then
+        it will use the decorated functions docstring and formats it.
+
+    args :
+        passed to :meth:`str.format`.
+
+    kwargs :
+        passed to :meth:`str.format`. If the function has a (not empty)
+        docstring the original docstring is added to the kwargs with the
+        keyword ``'__doc__'``.
+
+    Raises
+    ------
+    ValueError
+        If the ``docstring`` (or interpreted docstring if it was ``'__doc__'``
+        or not a string) is empty.
+
+    IndexError, KeyError
+        If a placeholder in the (interpreted) ``docstring`` was not filled. see
+        :meth:`str.format` for more information.
+
+    Notes
+    -----
+    Using this decorator allows, for example Sphinx, to parse the
+    correct docstring.
+
+    There might be problems with certain objects which have a not-writable
+    ``__doc__`` attribute. Like classes in python2.
+
+    Examples
+    --------
+
+    Replacing the current docstring is very easy::
+
+        >>> from astropy.utils.decorators import format_doc
+        >>> doc = '''Perform num1 + num2'''
+        >>> @format_doc(doc)
+        ... def add(num1, num2):
+        ...     return num1+num2
```

Comment:
I think for the examples I'd prefer that instead of using a separate `doc` variable first, just write:
```
@format_doc('''Perform num1 + num2''')
...
```
For longer docstrings it could still be split on multiple lines.  But there's no reason a multi-line docstring can't go in the format_doc call directly.  That's a highly subjective matter of taste though.  Maybe having one example of both approaches would be fine.